prusti-encoder/src/encoders/mir_impure.rs

```diff
@@ -802,22 +823,87 @@ impl<'vir, 'enc> mir::visit::Visitor<'vir> for EncoderVisitor<'vir, 'enc> {
                     _ => todo!(),
                 };
 
-                let func_out = self.deps.require_ref::<crate::encoders::MirImpureEncoder>(
-                    *func_def_id,
-                ).unwrap();
+                // TODO: dedup with mir_pure
+                let attrs = self.vcx.tcx.get_attrs_unchecked(*func_def_id);
+                let is_pure = attrs.iter()
+                    .filter(|attr| !attr.is_doc_comment())
+                    .map(|attr| attr.get_normal_item()).any(|item| 
+                        item.path.segments.len() == 2
+                        && item.path.segments[0].ident.as_str() == "prusti"
+                        && item.path.segments[1].ident.as_str() == "pure"
+                    );
+
+                let dest = self.encode_place(Place::from(*destination));
+                let call_args = args.iter().map(|op| 
+                    if is_pure {
+                        self.encode_operand_snap(op)
+                    }
+                    else {
+                        self.encode_operand(op)
+                    }
+                );
+
+                if is_pure {
+                    let func_args = call_args.collect::<Vec<_>>();
+                    let pure_func_name = self.deps.require_ref::<crate::encoders::MirFunctionEncoder>(
+                        *func_def_id,
+                    ).unwrap().function_name;
+
+                    let pure_func_app = self.vcx.mk_func_app(pure_func_name, &func_args);
+
+                    let method_assign = {
+                        //TODO: Can we get `method_assign` in a better way? Maybe from the MirFunctionEncoder?
+                        let body = self.vcx.body.borrow_mut().get_impure_fn_body_identity(func_def_id.expect_local());
+                        let return_type = self.deps
+                            .require_ref::<crate::encoders::TypeEncoder>(body.return_ty())
+                            .unwrap();
+                        return_type.method_assign
```

Comment:
It would be better to just extend `MirFunctionEncoder` to also return the `return` `vir::Type` since it will already have it available instead of going through the convoluted `self.vcx.body.borrow_mut().get_impure_fn_body_identity`.